src/serializer/cookie_utils.rs

```diff
@@ -0,0 +1,206 @@
+use cookie_factory::bytes::{be_f64, be_i16, be_i8};
+use cookie_factory::combinator::{skip, slice};
+use cookie_factory::gen::legacy_wrap;
+use cookie_factory::GenError;
+
+pub(crate) fn gen_at_offset<G>(
+    offset: usize,
+    f: G,
+) -> impl Fn((&mut [u8], usize)) -> Result<(&mut [u8], usize), GenError>
+where
+    G: Fn((&mut [u8], usize)) -> Result<(&mut [u8], usize), GenError>,
+{
+    move |input| {
+        let old_pos = input.1;
+        match input.0.len() < offset {
+            false => f((input.0, offset)).map(|(input, _)| (input, old_pos)),
+            true => Err(GenError::BufferTooSmall(offset - input.0.len())),
+        }
+    }
+}
+
+pub(crate) fn gen_opt<'a, T, G, H>(
+    val: Option<&'a T>,
+    f: G,
+) -> impl Fn((&'a mut [u8], usize)) -> Result<(&'a mut [u8], usize), GenError>
+where
+    G: Fn(&'a T) -> H,
+    H: Fn((&'a mut [u8], usize)) -> Result<(&'a mut [u8], usize), GenError>,
+{
+    move |input| {
+        if let Some(val) = val {
+            f(val)(input)
+        } else {
+            Ok(input)
+        }
+    }
+}
+
+pub(crate) fn gen_opt_copy<'a, T: Copy, G, H>(
+    val: Option<T>,
+    f: G,
+) -> impl Fn((&'a mut [u8], usize)) -> Result<(&'a mut [u8], usize), GenError>
+where
+    G: Fn(T) -> H,
+    H: Fn((&'a mut [u8], usize)) -> Result<(&'a mut [u8], usize), GenError>,
+{
+    move |input| {
+        if let Some(val) = val {
+            f(val)(input)
+        } else {
+            Ok(input)
+        }
+    }
+}
+
+pub(crate) fn gen_many<'a, T: IntoIterator + Copy, G, H>(
+    list: T,
+    f: G,
+) -> impl Fn((&'a mut [u8], usize)) -> Result<(&'a mut [u8], usize), GenError>
+where
+    G: Fn(<T as IntoIterator>::Item) -> H,
+    H: Fn((&'a mut [u8], usize)) -> Result<(&'a mut [u8], usize), GenError>,
+{
+    move |input| {
+        list.into_iter().fold(Ok(input), |r, v| match r {
+            Err(e) => Err(e),
+            Ok(x) => f(v)(x),
+        })
+    }
+}
+
+pub(crate) fn tuple<'a, List: tuple::Tuple<'a>>(
```

Comment:
How does this differ from the `tuple` from cookie factory?

Reply:
`tuple` from cookie factory does not implement the `legacy_wrap` function, so we had to crate our own here